drivers/hwinfo/hwinfo_litex.c

```diff
@@ -0,0 +1,24 @@
+/*
+ * Copyright (c) 2019 Antmicro <www.antmicro.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <drivers/hwinfo.h>
+#include <soc.h>
+#include <string.h>
+#include <device.h>
+#include <sys/util.h>
+
+ssize_t z_impl_hwinfo_get_device_id(u8_t *buffer, size_t length)
+{
+	u8_t volatile *ptr = (u8_t volatile *)(DT_INST_0_LITEX_DNA0_BASE_ADDRESS);
+	u8_t end = MIN(length, (DT_INST_0_LITEX_DNA0_SIZE / 8));
+
+	for (int i = 0; i < end; i++) {
+		buffer[i] = *(ptr + (i * 0x04)); /* In each CSR only lower 8 bits are used */
```

Comment:
Can you point me to the documentation where I can find some information about this?

Reply:
I believe that in case of LiteX, the source code together with internal comments *is* the documentation. 

The DNA's code is here: https://github.com/enjoy-digital/litex/blob/master/litex/soc/cores/dna.py, but you are most probably interested in the CSRStatus: https://github.com/enjoy-digital/litex/blob/master/litex/soc/interconnect/csr.py#L125.

Reply:
I was asking for documentation about this DNA module — something about the content of this constants.
The comment from the DT defines this register as 
"DNA value is 47-bits long, register width is aligned to 64",
But here you have ` *(ptr + (i * 0x04));` where `i` is maximum 8 (DT_INST_0_LITEX_DNA0_SIZE / 8).
`8 * 0x04` evaluates to 32. That seems like an overrun, right?
The code is the same as:
```C
u32_t volatile *ptr = (u32_t volatile *)(DT_INST_0_LITEX_DNA0_BASE_ADDRESS);
ssize_t end = MIN(length, (DT_INST_0_LITEX_DNA0_SIZE / 8));

for (int i = 0; i < end; i++) {
	buffer[i] = ptr[i] & 0x08; /* In each CSR only lower 8 bits are used */
}
```

Reply:
I found another source that might shed some light on how CSRs are defined in LiteX: https://github.com/timvideos/litex-buildenv/wiki/LiteX-for-Hardware-Engineers#csrs-config-and-status-registers. The most important part is:
```
CSRs are a bit odd, by default they are byte-wide registers that are
on 32-bit word boundaries.
So a "32-bit" CSR is actually broken into four bytes spanning a total
address space of 16 bytes.
```
I've just noticed there is a typo in the comment that is quite importaint in the discussion -- DNA ID has **57** bits instead of 47. It means that 8 bytes are required to represent it (as 57 = 7*8 + 1).

As it does not fit in a single 32-bit register, it is exposed as a 64-bit CSR (which can be treated as two 32-bit CSRs). Since LiteX generates a single 32-bit CSR as 4 consecutive 32-bit registers, there are 8 registers in total (each one containing meaningful data stored in the least-significant byte).

I'll fix the typo and rephrase the comment in the DTS to make things clearer.

Reply:
57 comes from: https://github.com/enjoy-digital/litex/blob/master/litex/soc/cores/dna.py#L11  which corresponds to the info on `DNA_PORT` on page 111 in this [document](https://www.xilinx.com/support/documentation/sw_manuals/xilinx14_7/7series_hdl.pdf).

Reply:
Can we use the code from my snipped above? I think that notation is somehow more clear.

Reply:
Yes. I am reprhasing the comment in the code as well and I think it will fit better in the context of your notation.
I think thougt, that it should be: `buffer[i] = ptr[i] & 0xFF`. Plus I would add casting.

Reply:
You are right. Don't know why I wrote 0x08. An explicit cast is always a good idea.